### PR TITLE
Added a TextureManager Map to store a TextureManager for each GraphicCon...

### PR DESCRIPTION
**Pull request by jtorresfabra**
...text. This way we can emulate the behaviour of OSG when releasing textures. 

It is not a complete fix for all the releaseGL stuff, it's only for Textures so DatabasePager frees Texture memory again.  
We can discuss it further, but I think this way we can release other GLObjects with and without state, as OSG does. In OSG a static variable keeps track of GLObjects or each class with respect a contextID.

### CODE REVIEW

**Comment by jtorresfabra** [comment]
Should fix #291 #222.

File: sources/osg/StateSet.js [hunk 1]
```diff
@@ -217,6 +217,12 @@ define( [
             for ( var i = 0, j = this.textureAttributeMapList.length; i < j; i++ ) {
                 this.getTextureAttribute( i, 'Texture' ).releaseGLObjects( state );
             }
+            var list = this.getAttributeList();
+                for ( i = 0, j = list.length; i < j; i++ ) {
```
Comment (cedricpinson) [inline sources/osg/StateSet.js]: indentation seems bad

**Comment by cedricpinson** [comment]
I dont see method to mark data to release, I expected to see something like:
texture.markToRelease() or someting like that. Just to be sure it 's something to add in the futur right ?

File: sources/osg/StateSet.js (hunk 1 above)
Comment (jtorresfabra) [inline sources/osg/StateSet.js]: Fixed

**Comment by jtorresfabra** [comment]
Nope, you simply need to call texture.releaseGLObjects() to get it deleted asynchronously in the next frame. If you want to delete it from a specific Context you pass the state as a parameter, if you don't pass the state then it will be deleted from all contexts. 

Ping me if you need a better explanation, maybe some comments in the code could be good.

**Comment by cedricpinson** [comment]
I understand and maybe we dont need more, but in c++ you have destructor to tell a object to release, but not in js. I thought that it would be helpful to have a special function to say release this texture, but without more arguments. Because the texture is created from a specific context the texture is able to release from the same context. I guess that seems we dont have shared resources between context, the no arguments that delete from all context does the job.
So maybe we dont need more function to do that for now

**Comment by jtorresfabra** [comment]
>I understand and maybe we dont need more, but in c++ you have destructor to tell a object to release, but not in js. I thought that it would be helpful to have a special function to say release this texture, but without more arguments. Because the texture is created from a specific context the texture is able to release from the same context.

Yes, but a texture can be used in several contexts, ( maybe not now in osgjs, but it will/should do) so you need to specify from which context you want to release or release from all contexts (as it works in OSG for Textures and other GLObjects).

>I guess that seems we dont have shared resources between context, the no arguments that delete from all context does the job.

Yes that's the point. 

>For example if you want to release all resource of Geometry you can't call the function from from geometry, you need state. So it does not emulate a destructor.

Now BufferArray is missing a BufferObjectManager (as Texture has its own TextureManager) to be able to release from all contexts. In fact OSG has a BufferObjectManager. I implemented the BufferArray static to delete cache the easy way, not using a manager, I tried to say it in the IRC. I wanted to show the idea, and how it works.
 
Once we have this manager you will be able to call releaseGLObjects on Geometry safely without passing the state. The most internal classes (TextureManager and BufferArrayManager) will be responsibles of managing the "no-state" situation, as TextureManager does now. So for Geometry class it will be trasnparent.  

For Shaders and Programs it is not the same, in OSG you can only release them using a context ID, not as BufferObejcts and Textures, which makes sense.

**Comment by jtorresfabra** [comment]
I've been having a look to the "Sharing WebGLresources" state. It seems is a proposal/draft https://www.khronos.org/webgl/wiki/SharedResouces of extension. Unfortunately it seems latest chrome does not support the extension (at least mine).
https://www.khronos.org/registry/webgl/sdk/tests/conformance/context/resource-sharing-test.html?webglVersion=1

Said that, TextureManager is prepared in case we need that support. It won't fail trying to delete a texture from a different context from where it was created. 
For Buffer Objects/Programs/Shaders we can simply keep a reference to the glcontext when the resource is binded/created and store it as a class variable. Then you can call releaseGLObjects() without the state as a paramenter. The real thing we need to delete resources is the GL context, not the state. Implement that based on this PR would be pretty straight forward. 

Thoughts?

**Comment by jtorresfabra** [comment]
@cedricpinson:

After looking at http://webglstats.com to check the availability of WebGL Shared Resources extensions. I've just gone ahead with the idea of having a reference to the graphic context in each glresource and it seems to work well. Now you don't need to call releaseGLResources with a parameter. So at the end calling releaseGLResources() is like doing that markToBeReleased we were talking.

**Comment by cedricpinson** [comment]
Ok,
the only thing missing is the manager for other GL resources to have the same way to delete textures, right ?

**Comment by jtorresfabra** [comment]
Really we don't need it. The textures stuff is different because it is using a TextureManager to reuse the texture objects.

**Comment by cedricpinson** [comment]
ok but it means you dont control the time / delay to delete those resource because it's not delayed by a manager, and applied per frame. Or maybe I miss something.

**Comment by jtorresfabra** [comment]
Indeed it IS controlled. Each GLResource has its own static to delete cache, which is flushed every frame.
For example (For buffer array):
When ther is a call to releaseGLObejcts() the buffer is pushed to the cache.
https://github.com/cedricpinson/osgjs/pull/297/files?diff=unified#diff-e466d62e2e95574ba588e47efa37270aR50
Then the next frame the viewer flushes that cache:
https://github.com/cedricpinson/osgjs/pull/297/files?diff=unified#diff-50f9ea5b8e120acc5d8e00a8ae9cb389L275

**Comment by jtorresfabra** [comment]
Maybe it could be easier to understand loading the PagedLOD example a putting a breakpoint in  the above lines. Then play a little with the example and it will be stopped at the breakpoints deleting buffers.

**Comment by jtorresfabra** [comment]
Updated the PR as discussed in IRC. It's ready for review ( I cannot change the PR labels ).

**Pull request merged**